source/tutorials/nixos-configuration-testing.md

```diff
@@ -0,0 +1,485 @@
+    in ''
+      machine.wait_until_succeeds("${wmClass} | grep -q 'firefox'")
+      machine.sleep(20)
+      machine.screenshot("screen")
+    '';
+```
+
+The complete `firefoxtest.nix` file looks like the following:
+```{code-block}
+let
+  nixpkgs = builtins.fetchTarball "https://github.com/nixOS/nixpkgs/archive/22.05.tar.gz";
+  pkgs = import nixpkgs {};
+  home-manager = builtins.fetchTarball "https://github.com/nix-community/home-manager/archive/release-22.05.tar.gz";
+in
+  pkgs.nixosTest {
+    nodes.machine = {...}: {
+      imports = [
+        (import "${home-manager}/nixos")
+      ];
+      boot.loader.systemd-boot.enable = true;
+      boot.loader.efi.canTouchEfiVariables = true;
+
+      services.xserver.enable = true;
+      services.xserver.displayManager.gdm.enable = true;
+      services.xserver.desktopManager.gnome.enable = true;
+      services.xserver.displayManager.autoLogin.enable = true;
+      services.xserver.displayManager.autoLogin.user = "alice";
+
+      users.users.alice = {
+        isNormalUser = true;
+        extraGroups = ["wheel"]; # Enable ‘sudo’ for the user.
+        uid = 1000;
+      };
+
+      home-manager.users.alice = {
+        home.packages = [
+          pkgs.firefox
+          pkgs.thunderbird
+        ];
+      };
+
+      system.stateVersion = "22.05";
+
+      environment.systemPackages = [
+        (pkgs.makeAutostartItem {
+          name = "firefox";
+          package = pkgs.firefox;
+        })
+      ];
+
+      systemd.user.services = {
+        "org.gnome.Shell@wayland" = {
+          serviceConfig = {
+            ExecStart = [
+              # Clear the list before overriding it.
+              ""
+              # Eval API is now internal so Shell needs to run in unsafe mode.
+              "${pkgs.gnome.gnome-shell}/bin/gnome-shell --unsafe-mode"
+            ];
+          };
+        };
+      };
+    };
+
+    testScript = {nodes, ...}: let
+      user = nodes.machine.config.users.users.alice;
+      #uid = toString user.uid;
+      bus = "DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/${toString user.uid}/bus";
+      gdbus = "${bus} gdbus";
+      su = command: "su - ${user.name} -c '${command}'";
+      gseval = "call --session -d org.gnome.Shell -o /org/gnome/Shell -m org.gnome.Shell.Eval";
+      wmClass = su "${gdbus} ${gseval} global.display.focus_window.wm_class";
+    in ''
+      machine.wait_until_succeeds("${wmClass} | grep -q 'firefox'")
+      machine.sleep(20)
+      machine.screenshot("screen")
+    '';
+  }
+```
+
+### Tests that need multiple virtual machines
+
+Tests can utilize multiple virtual machines.
+As an example we use the use-case of a REST interface to a Postgres database.
+Because some of the needed packages of this example are broken in 22.05 release this example uses a specific revision of nixpkgs.
+This example shows the value of pinning a test to a specific revision of `nixpkgs`.
```

Comment:
This should be a `:::{note}` right before the example. It is not well-placed in the introduction to what will come next, and really is a stupid implementation detail we'd rather not care about.